agent.go

```diff
@@ -768,9 +770,80 @@ func (a *Agent) handleNewPeerReflexiveCandidate(local *Candidate, remote net.Add
 	return nil
 }
 
+func (a *Agent) assertInboundUsername(m *stun.Message) error {
+	usernameAttr := &stun.Username{}
+	usernameRawAttr, usernameFound := m.GetOneAttribute(stun.AttrUsername)
+
+	if !usernameFound {
+		return fmt.Errorf("inbound packet missing Username")
+	} else if err := usernameAttr.Unpack(m, usernameRawAttr); err != nil {
+		return err
+	}
+
+	expectedUsername := a.localUfrag + ":" + a.remoteUfrag
+	if usernameAttr.Username != expectedUsername {
+		return fmt.Errorf("MessageIntegrity mismatch expected(%x) actual(%x)", expectedUsername, usernameAttr.Username)
+	}
+
+	return nil
+}
+
+func (a *Agent) assertInboundMessageIntegrity(m *stun.Message, key []byte) error {
+	messageIntegrityAttr := &stun.MessageIntegrity{}
+	messageIntegrityRawAttr, messageIntegrityAttrFound := m.GetOneAttribute(stun.AttrMessageIntegrity)
+
+	if !messageIntegrityAttrFound {
+		return fmt.Errorf("inbound packet missing MessageIntegrity")
+	} else if err := messageIntegrityAttr.Unpack(m, messageIntegrityRawAttr); err != nil {
+		return err
+	}
+
+	tailLength := 24
+	rawCopy := make([]byte, len(m.Raw))
+	copy(rawCopy, m.Raw)
+
+	if _, hasFingerprint := m.GetOneAttribute(stun.AttrFingerprint); hasFingerprint {
+		tailLength += 8
+
+		currLength := binary.BigEndian.Uint16(rawCopy[2:4])
```

Comment:
Can I see a diagram of this structure or something? Kind of mysterious what values  you're pulling out here.

Reply:
Updated this with comments,  just rewriting the STUN length header. I used a bunch of named variables (instead of magic numbers) hopefully a little better. If you still think it is bad tell me and can fix up